Source/AwsCommonRuntimeKit/crt/ShutdownCallbackCore.swift

```diff
@@ -0,0 +1,83 @@
+//  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//  SPDX-License-Identifier: Apache-2.0.
+
+import AwsCCommon
+import AwsCAuth
+
+public typealias ShutdownCallback = () -> Void
+
+/// Core classes have manual memory management.
+/// You have to balance the retain & release calls in all cases to avoid leaking memory.
+class ShutdownCallbackCore {
+    let shutdownCallback: ShutdownCallback
+    init?(_ shutdownCallback: ShutdownCallback?) {
+        guard let shutdownCallback = shutdownCallback else {
+            return nil
+        }
+        self.shutdownCallback = shutdownCallback
+    }
+
+    /// Calling this function performs a manual retain on the ShutdownCallbackCore.
+    /// and returns aws_shutdown_callback_options. When the shutdown finally fires,
+    /// it will manually release ShutdownCallbackCore.
+    ///
+    /// If you fail to create something that uses the aws_shutdown_callback_options,
+    /// you must call release() to avoid leaking memory.
+    func getRetainedShutdownOptions() -> aws_shutdown_callback_options {
+        var shutdown_options = aws_shutdown_callback_options()
+        shutdown_options.shutdown_callback_fn = { rawValue in
+            guard let rawValue = rawValue else {
+                return
```

Comment:
We should not guard this. We always pass user_data to this callback. If this is null, we should NOT silently ignore the error, we should terminate the application

Reply:
Thanks, fixed.

Reply:
is it something that needs to be asserted on? Sounds like a developer error?

Reply:
I'm ambivalent (leaning towards no) about adding additional asserts.

yes, this would be a developer error if it ever occurred. When we pass a `void *user_data*` pointer down to C, it is 100% expected that the same value comes back, and we ALWAYS pass a valid value here, so we never expect this to be NULL.

in the latest version it would terminate due to `userData!` so an assert wouldn't help with correctness, it would just help users understand the error if it ever occurred. but It's verrrrry unlikely that this ever occurs.